### ISSUE TITLE
Python3 issue with dbm

### ISSUE THREAD

**Issue by pyhalov** [avahi/avahi#260]
https://github.com/lathiat/avahi/commit/63750f1be96ad08c407193b08bf3b9ee74310e2d moved build-db and ServiceTypeDatabase.py to Python3 dbm, but python 3 dbm will always select gdbm over ndbm. This breaks --disable-gdbm --enable-dbm build.

**Comment by lathiat** [comment]
Hi @pyhalov thanks for the report. Seems we need to try and standardize on DBM in some way. Can you tell me your specific usecase for --disable-gdbm / --enable-dbm ?

**Comment by pyhalov** [comment]
I'm not sure, perhaps, just a bit of history - https://lists.freedesktop.org/archives/avahi/2006-August/000840.html

**Comment by lathiat** [comment]
Are you actually using solaris?

**Comment by lathiat** [comment]
Also does solaris now owrk if you just use the default, e.g. gdbm? I would think for modern solaris this would work.

**Comment by pyhalov** [comment]
No, but I support avahi on OpenIndiana and I'm a bit concerned about consequences (possible incompatibilities) for our users if we rebuild it with gdbm.

**Comment by lathiat** [comment]
Ok perfect thanks I’ll see what I can do here

**Comment by smcv** [comment]
Relatedly, since 63750f1, if Avahi is compiled for Python **2**, the service type database will preferentially be compiled with dbhash (bsddb), which is neither GNU gdbm nor traditional Unix (n)dbm, and can't necessarily be loaded by the other database libraries.

I ran across this because dbhash doesn't support looking up Unicode keys, which avahi-discover relies on (it appears they're transparently encoded to bytes by the other backends). The C code might also fail to look up service types if the database is in an unexpected format.

I think the solution is for this script to force a particular database backend, so that the service type database is always compiled to the expected format. I'll send a PR.

(With hindsight, this database could probably have been a simple text file on disk and a hash table in memory, since it's not exactly large... but that would be an incompatible change now.)

**Comment by evverx** [comment]
Looks like it got worse. Python3.13 switched the default backend to `sqlite3` (https://docs.python.org/3/whatsnew/3.13.html) so `build-db` builds it with the default backend and then, for example, `avahi-browse -b` can't read it.

**Comment by pemensik** [comment]
It seems dbm interface does not even allow us to choose gdbm as preferred format. The problem with this is avahi-browse tries to use the same file too, but does not follow the same changes as python dbm interface.

That makes dbm interface quite a bad choice to create that file. I think instead we should implement these lookups in native and export them into python too.

**Comment by evverx** [comment]
FWIW `--enable-dbm --disable-gdbm` has been broken for some time apparently (regardless of https://github.com/avahi/avahi/commit/9fde934cd449a7ac99c0c32424b1c3c5cfd7f50e). I opened https://github.com/avahi/avahi/issues/738 but it seems it can safely be dropped. I'd assume it was reported if it was actually used anywhere.

**Comment by evverx** [comment]
OpenIndiana apparently passes `--enable-dbm --disable-gdbm` in https://github.com/OpenIndiana/oi-userland/blob/44d8f6081b1773abdd8343a9e12c8fbd8aa42d10/components/network/avahi/Makefile#L62-L63. On the other hand it also carries patches like https://github.com/OpenIndiana/oi-userland/blob/oi/hipster/components/network/avahi/patches/27-no-svc-type-database.patch where `service-type-database` is removed. I'm not sure what's going on there but given that it seems to be the only place where it was ever used and broken everywhere else it can probably be removed upstream and moved downstream if necessary. @klausz65 I wonder if it would be OK to remove the `enable-dbm` part of avahi?

**Comment by klausz65** [comment]
Okay, I removed both options: --disable-gdbm and --enable-dbm. See: https://github.com/OpenIndiana/oi-userland/pull/24599

**Comment by evverx** [comment]
Thank you!

I was planning to open a PR removing the dbm support but it seems the Fedora package breaks somehow: https://download.copr.fedorainfracloud.org/results/packit/evverx-avahi-3/fedora-rawhide-x86_64/09785015-avahi/builder-live.log.gz
```sh
RPM build errors:
    Directory not found: /builddir/build/BUILD/avahi-0.8-build/BUILDROOT/usr/lib64/avahi
Finish: rpmbuild avahi-0.8-6.20251110234428993360.pr3.320.g685d612.fc44.src.rpm
```
The Fedora package does pass nonsensical flags (`--disable-python --enable-gdbm`) but those flags doesn't break anything else. @pemensik do you know why the Fedora package somehow breaks when just the following part is removed
https://github.com/avahi/avahi/blob/43cabe1e137a67d0ce23e192c1fdbd462c93a4e1/service-type-database/Makefile.am#L37-L52
and where that error comes from?

**Comment by evverx** [comment]
Looks like it's the broken `WITH_PYTHON` thing in the Fedora package again. I got it to work by removing `%dir %{_libdir}/avahi` from https://src.fedoraproject.org/rpms/avahi/blob/rawhide/f/avahi.spec#_638. @pemensik shouldn't it go to `WITH_PYTHON`? Either way I'll open a separate issue about `WITH_PYTHON`. It's technically a downstream bug but since Packit is used upstream to test avahi it should be addressed.
